File: 200. Number of Islands/numIslands.java
class Solution {
    public int numIslands(char[][] grid) {
        if (grid == null || grid.length == 0) {
            return 0;
        }

        int width = grid.length;
        int length = grid[0].length;
        int cnt = 0;
        for (int r = 0; r < width; ++r) {
            for (int c = 0; c < length; ++c) {
                if (grid[r][c] == '1') {
                    ++cnt;
                    dfs(grid, r, c);
                }
            }
        }

        return cnt;
    }

    public void dfs(char[][] grid, int r, int c) {
        int width = grid.length;
        int length = grid[0].length;

        if (r < 0 || c < 0 || r >= width || c >= length || grid[r][c] == '0') {
            return;
        }

        grid[r][c] = '0';
        dfs(grid, r - 1, c);
        dfs(grid, r + 1, c);
        dfs(grid, r, c - 1);
        dfs(grid, r, c + 1);
    }
}

File: 200. Number of Islands/numIslands.py
import queue
class numIslands:
    def solution(self, grid: List[List[str]]) -> int:
        width = len(grid)
        if width == 0:
            return 0
        length = len(grid[0])
        cnt = 0
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        q = queue.Queue()
        for i in range(width):
            for j in range(length):
                if grid[i][j] == '1':
                    q.put((i, j))
                    cnt = cnt + 1
                    while not q.empty():
                        temp = q.get()
                        for d in directions:
                            x = temp[0] + d[0]
                            y = temp[1] + d[1]
                            if x >= 0 and x < width and y >= 0 and y < length and grid[x][y] == '1':
                                grid[x][y] = '0'
                                q.put((x, y))
        return cnt